java/arcs/sdk/Particle.kt

```diff
@@ -11,8 +11,33 @@
 
 package arcs.sdk
 
+/**
+ * Interface used by [ArcHost]s to interact dynamically with code-generated [Handle] fields
+ * used by [Particle]s.
+ */
+interface HandleHolder {
+    val map: MutableMap<String, Handle>
+    val entitySpecs: MutableMap<String, EntitySpec<out Entity>>
+}
+
+/**
+ * Base class used by `schema2kotlin` code-generator tool to generate a class containing all
+ * declared handles.
+ */
+abstract class HandleHolderBase(
```

Comment:
I don't think you need this class at all. The code generator can generate code that looks like this:

```
class FooHandles : HandleHolder {
    override val map = mapOf(
        "foo" to SingletonImpl(...),
        ...
    )
}
```

Then you don't need the empty mutable map to already exist.

Reply:
The code generator isn't allowed to instantiate these Handles, they must be instantiated by ArcHost via HandleManager, which is platform dependent.  Additionally, the StorageKeys are created by the Allocator, which doesn't happen until runtime. 

At best, these could be some kind of placeholder flyweights which are lazily initialized in place with real handles, but some kind of mutation is unavoidable.

Reply:
I think we can fix this in a different way by code-gening a method to create the handle classes and take the handle map and initialize the handles field, e.g.

abstract class BaseParticle<T> : Particle {
   <T : HandleHolder> setupHandles(map: Map<String, Handle>)
}

And then have the AbstractFooParticle<FooHandles> class code-gen this method. 

Then ArcHost can create all the handles, create an immutable map, and invoke BaseParticle.setupHandles(map)

Reply:
Making these truly immutable turned out to be more trouble than it was worth, so I instead simply declared them as Map<> but behind the scenes a MutableMap is used, which ArcHost will cast to. 

The only real way to achieve real immutability would be to make Particle ctor take these params, but we want to keep particle instantiation as arg-less ctors.

Reply:
Fair enough, thanks for investigating